File: nodehat/MotorSpeed.js
class PIController {
  constructor(
    proportional_constant = 0,
    integral_constant = 0,
    derivative_constant = 0
  ) {
    this.proportional_constant = proportional_constant;
    this.integral_constant = integral_constant;
    this.derivative_constant = derivative_constant;
    // Running sums
    this.integral_sum = 0;
    this.previous = 0;
  }

  handle_proportional(error) {
    return this.proportional_constant * error;
  }

  handle_integral(error) {
    this.integral_sum += error;
    return this.integral_constant * error;
  }

  handle_derivative(error) {
    const derivative = this.derivative_constant * (error - this.previous);
    this.previous = error;
    return derivative;
  }

  get_value(error) {
    let p = this.handle_proportional(error);
    let i = this.handle_integral(error);
    let d = this.handle_derivative(error);
    return p + i + d;
  }
}

class MotorSpeed {
  constructor(motor) {
    this.motor = motor;
    this.p = 0.01;
    this.k = 0;
    this.d = 0.02;
    this.pid = new PIController(this.p, this.k, this.d);
    //power should probably be passed by previouse behaviour
    this.power = 0;
    this.errors = [];
    this._setPoint = 0;
    this.running = false;
    this.update = this.update.bind(this);
  }

  update(data) {
    let { speed } = data;
    let error = speed - this.setPoint;
    //console.log("error", error);
    this.errors.push(error);
    let adjustment = this.pid.get_value(error);
    this.power = this.power - adjustment;
    console.log(speed, error, adjustment, this.power);
    this.motor.pwm = this.power;
  }

  set setPoint(speed) {
    if (speed < 0) {
      throw new Error("we havent dealt with going backwards yet");
    }
    this._setPoint = speed;
    console.log("set", this._setPoint);
  }

  get setPoint() {
    return this._setPoint;
  }

  start(setPoint) {
    if (setPoint < 0) {
      throw new Error("we havent dealt with going backwards yet");
    }
    this.setPoint = setPoint;
    console.log(this.setPoint);
    this.power = Math.max(this.motor.pwm, this.setPoint / 4);
    this.running = true;
    this.motor.pwm = this.power;
    this.motor.on("encoder", this.update);
  }

  stop() {
    console.log("in stop");
    this.motor.removeListener("encoder", this.update);
    this.running = false;
  }
}

module.exports = { MotorSpeed };

File: atomic_behaviours/MotorSpeed.py
from utilities.pid_controller import PIController
import json
import time
import pathlib
import os
import numpy as np
import matplotlib.pyplot as plt

print(pathlib.Path.cwd())

class MotorSpeed:
   def __init__(self,motor,set_point=40):
      self.motor = motor
      self.p = 0.01
      self.k = 0
      self.d = 0.02
      self.pid = PIController(self.p,self.k,self.d)
      ##power should probably be passed by previouse behaviour
      self.power=set_point/4
      self.errors = []
      self.set_point = set_point
      self.running = False
      self.manager = None

   def save_data(self):
      tune_data_folder = pathlib.Path.cwd().joinpath('tune_motor_speed')
      file_index = len(os.listdir(tune_data_folder))
      file_suffix = f'{file_index}_p-{self.p}_k-{self.k}_d-{self.d}'
      write_file_path = tune_data_folder.joinpath(f'{self.encoder.name}_{file_suffix}')
      json_path = f'{write_file_path}.json'
      png_path = f'{write_file_path}.png'
      with open(json_path,'w') as f:
         json.dump(
            {
            'p':self.p,
            'k':self.k,
            'd':self.d,
            'errors': self.errors,
            },f,indent=2)

      plt.plot(self.errors)
      plt.savefig(png_path)

   def transtion(self):
      return 
      if 'should transition':
         self.stop()
         self.manager.load_new_behaviour('some behaviour')

   async def update(self,data):
      _,speed,_,_,_ = data
      error = speed-self.set_point
      self.errors.append(error)
      adjustment = self.pid.get_value(error)
      self.power = self.power-adjustment
      self.motor.pwm(self.power)

   def start(self):
      self.running = True
      self.motor.pwm(self.power)
      self.motor.add_listener('encoder',self.update)

   def stop(self):
      self.motor.remove_listener('encoder',self.update)
      self.motor.pwm(0)
      #self.save_data()

   
      
